Data/Primitive/Array/Checked.hs

```diff
@@ -0,0 +1,225 @@
+{-# LANGUAGE TypeFamilies #-}
+
+{-|
+Module      : Data.Array.Checked
+Description : Bounded checked boxed and unboxed arrays
+Copyright   : (c) Winter, 2017
+License     : BSD3
+Maintainer  : libraries@haskell.org, drkoster@qq.com
+
+Stability   : experimental
+Portability : non-portable
+
+This module provides exactly the same API with "Data.Primitive.Array.Class", but will throw an 'IndexOutOfBounds'
+'ArrayException' on bound check failure.
+
+-}
+module Data.Primitive.Array.Checked
+  ( -- * Arr typeclass re-export
+    A.Arr
+  , RealWorld
+    -- * Bound checked array operations
+  , newArr
+  , newArrWith
+  , readArr
+  , writeArr
+  , setArr
+  , indexArr
+  , indexArrM
+  , freezeArr
+  , thawArr
+  , copyArr
+  , copyMutableArr
+  , moveArr
+  , cloneArr
+  , cloneMutableArr
+  , resizeMutableArr
+  , shrinkMutableArr
+  -- * No bound checked operations
+  , A.unsafeFreezeArr
+  , A.unsafeThawArr
+  , A.sameMutableArr
+  , A.sizeofArr
+  , A.sizeofMutableArr
+  , A.sameArr
+  -- * Boxed array type
+  , A.Array(..)
+  , A.MutableArray(..)
+  , A.SmallArray(..)
+  , A.SmallMutableArray(..)
+  , A.uninitialized
+  -- * Primitive array type
+  , A.PrimArray(..)
+  , A.MutablePrimArray(..)
+  , A.newPinnedPrimArray, A.newAlignedPinnedPrimArray
+  -- * Bound checked primitive array operations
+  , copyPrimArrayToPtr, copyMutablePrimArrayToPtr, copyMutablePrimArrayFromPtr
+  -- * Unlifted array type
+  , A.UnliftedArray(..)
+  , A.MutableUnliftedArray(..)
+  , A.PrimUnlifted(..)
+  -- * The 'ArrayException' type
+  , ArrayException(..)
+  ) where
+
+import qualified Data.Primitive.Array.Class as A
+import Control.Exception (throw, ArrayException(..))
+import Control.Monad.Primitive
+import Data.Primitive.Types
+import GHC.Ptr (Ptr(..))
+
+check :: String -> Bool -> a -> a
+check _      True  x = x
+check errMsg False _ = throw (IndexOutOfBounds $ "Data.Array.Checked." ++ errMsg)
+{-# INLINE check #-}
```

Comment:
for this module, does INLINE pragmas make sense? theres definitely no fusion going on , and to quote @thoughtpolice on the topic, if ghc isn't marking them inlinable when they're small enough, thats a ghc bug ..